### PR TITLE
SWATCH-2465: Do not filter by metric if not set in the Instances API

### PR DESCRIPTION
**Pull request by Sgitario**
Jira issue: [SWATCH-2465](https://issues.redhat.com/browse/SWATCH-2465)

## Description
Before, we were using the default (first) metric ID by product tag if the user does not specify any. After these changes, we'll be getting all the metrics if the user does not specify any.  

To avoid getting duplicates, I had to update the "tally_instance_view" to aggregate all the metrics in the host into a single record. 

## Testing
IQE Test MR: https://gitlab.cee.redhat.com/insights-qe/iqe-rhsm-subscriptions-plugin/-/merge_requests/664

### Steps

1.- podman-compose up
2.- Add host to Insight DB:

```
INSERT INTO hosts (id, account, display_name, created_on, modified_on, facts, tags, canonical_facts, system_profile_facts, ansible_host, stale_timestamp, reporter, per_reporter_staleness, org_id, "groups") 
VALUES ('e5406d59-5cf3-4e83-abf8-84c149f8b6ba', 'account123', 'd125f119-0c98-43f2-a5de-ccf2e8e1ab55', '1993-03-26 00:00:00.000000 +00:00', '1993-03-26 00:00:00.000000 +00:00', '{"rhsm": {"orgId": "org123", "IS_VIRTUAL": null, "VM_HOST_UUID": null}}', null, '{"insights_id": "d125f119-0c98-43f2-a5de-ccf2e8e1ab55", "subscription_manager_id": "ef5a9896-242c-44a4-908d-9e6a1ffb9df6", "provider_id": "d125f119"}', '{"host_type": null, "cloud_provider": null, "is_marketplace": false, "cores_per_socket": 4, "number_of_sockets": 1, "installed_products": [{"id": "290"}], "infrastructure_type": "PHYSICAL"}', null, '2030-01-01 00:00:00.000000 +00:00', 'rhsm-conduit', '{}', 'org123', '{}');
```

The important part of the above query is `installed_products` which contains the value "290" to use the product "OpenShift Container Platform". This product will return both Sockets and Cores metrics.

3.- DEV_MODE=true ./gradlew :bootRun
4.- Run nightly Tally using command below:

```
http PUT ":8000/api/rhsm-subscriptions/v1/internal/rpc/tally/snapshots/org123" \
Origin:console.redhat.com \
x-rh-swatch-psk:placeholder
```

5.- Verification in RHSM subscription database:

```
echo -n '{"identity":{"account_number":"","type":"User","user":{"is_org_admin":true},"internal":{"org_id":"org123"}}}' | base64 -w 0
-- eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6IiIsInR5cGUiOiJVc2VyIiwidXNlciI6eyJpc19vcmdfYWRtaW4iOnRydWV9LCJpbnRlcm5hbCI6eyJvcmdfaWQiOiJvcmcxMjMifX19
```

```
curl -X 'PUT' \
  'http://localhost:8000/api/rhsm-subscriptions/v1/internal/rpc/tally/opt-in?org_id=org123' \
  -H 'accept: application/vnd.api+json' \
  -H 'x-rh-identity: eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6IiIsInR5cGUiOiJVc2VyIiwidXNlciI6eyJpc19vcmdfYWRtaW4iOnRydWV9LCJpbnRlcm5hbCI6eyJvcmdfaWQiOiJvcmcxMjMifX19'
```

```
http GET ":8000/api/rhsm-subscriptions/v1/instances/products/OpenShift Container Platform" x-rh-identity:eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6IiIsInR5cGUiOiJVc2VyIiwidXNlciI6eyJpc19vcmdfYWRtaW4iOnRydWV9LCJpbnRlcm5hbCI6eyJvcmdfaWQiOiJvcmcxMjMifX19 | jq
```

It should return:

```json
{
  "data": [
    {
      "id": "6529c337-8b2f-48eb-8ed5-b064c725822e",
      "instance_id": "d125f119",
      "display_name": "d125f119-0c98-43f2-a5de-ccf2e8e1ab55",
      "measurements": [
        4.0,
        2.0
      ],
      "last_seen": "1993-03-26T00:00:00Z",
      "number_of_guests": 0,
      "category": "physical",
      "subscription_manager_id": "ef5a9896-242c-44a4-908d-9e6a1ffb9df6",
      "inventory_id": "e5406d59-5cf3-4e83-abf8-84c149f8b6ba"
    }
  ],
  "meta": {
    "count": 1,
    "product": "OpenShift Container Platform",
    "measurements": [
      "Cores",
      "Sockets"
    ]
  }
}
```

### CODE REVIEW

File: swatch-core/src/main/java/org/candlepin/subscriptions/db/RhsmSubscriptionsDataSourceConfiguration.java [hunk 1]
```diff
@@ -80,11 +84,14 @@ public HikariDataSource rhsmSubscriptionsDataSource(
...
   public LocalContainerEntityManagerFactoryBean rhsmSubscriptionsEntityManagerFactory(
       EntityManagerFactoryBuilder builder,
+      ConfigurableListableBeanFactory beanFactory,
       @Qualifier("rhsmSubscriptionsDataSource") DataSource dataSource) {
     return builder
         .dataSource(dataSource)
         .packages("org.candlepin.subscriptions.db.model")
         .persistenceUnit("rhsm-subscriptions")
+        .properties(
+            Map.of(ManagedBeanSettings.BEAN_CONTAINER, new SpringBeanContainer(beanFactory)))
```
Comment (awood) [inline swatch-core/src/main/java/org/candlepin/subscriptions/db/RhsmSubscriptionsDataSourceConfiguration.java:94]: What does this change do (out of curiosity)?

File: swatch-core/src/main/java/org/candlepin/subscriptions/db/model/converters/TallyInstanceViewMetricsConverter.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      throw new IllegalArgumentException("Error parsing metrics", e);
+    }
+  }
+
+  private double toDouble(Object o) {
+    if (o instanceof Integer i) {
+      return i;
+    } else if (o instanceof Double d) {
+      return d;
+    } else if (o instanceof Long l) {
```
Comment (awood) [inline swatch-core/src/main/java/org/candlepin/subscriptions/db/model/converters/TallyInstanceViewMetricsConverter.java:81]: Might want to handle `float` just to cover all the bases.

**Review by awood — APPROVED** [review]

File: swatch-core/src/main/java/org/candlepin/subscriptions/db/RhsmSubscriptionsDataSourceConfiguration.java (hunk 1 above)
Comment (Sgitario) [reply swatch-core/src/main/java/org/candlepin/subscriptions/db/RhsmSubscriptionsDataSourceConfiguration.java:94]: This is to inject beans into the attribute converters. Thanks to this property, I could remove the workaround in the event record converter (and use the injection in the new converter I added as part of this pull request): https://github.com/RedHatInsights/rhsm-subscriptions/pull/3291/files#diff-2eb15a053fce1dd8e73d4d07bc84d3929489d972aaf93c5cfb207fc63408c320L40

I will add a comment to state this as well.

File: swatch-core/src/main/java/org/candlepin/subscriptions/db/model/converters/TallyInstanceViewMetricsConverter.java (hunk 2 above)
Comment (Sgitario) [reply swatch-core/src/main/java/org/candlepin/subscriptions/db/model/converters/TallyInstanceViewMetricsConverter.java:81]: good idea. adding it.

**Pull request merged**